Webapp/sources/services/summary.py

```diff
@@ -0,0 +1,317 @@
+    """
+    solvent_data_keys = [
+        "solventTableNumbers",
+        "solvents",
+        "solventVolumes",
+        "solventHazards",
+        "solventPhysicalForms",
+    ]
+    solvent_dict = get_request_data_from_keys(solvent_data_keys)
+    services.hazard_code.get_multiple_compounds_data(solvent_dict, "solvent")
+
+    # Concurrency for if no solvents are chosen
+    solvent_dict["number_of_solvents"] = request.form["numberOfSolvents"]
+    if solvent_dict["number_of_solvents"] == "0":  # if no solvents have been chosen
+        solvent_dict["number_of_solvents"] = 1  # then it shows only one empty cell
+        solvent_dict["solvents"] = [" "]
+
+    # handle solvent primary keys
+    solvent_primary_keys_ls = auxiliary.get_data("solventPrimaryKeys")
+    solvent_dict["solvent_primary_keys_str"] = ", ".join(solvent_primary_keys_ls)
+    solvent_dict["solvent_primary_keys_list"] = solvent_primary_keys_ls
+    # get the smiles from the primary keys
+    solvent_dict["solvent_smiles_ls"] = services.all_compounds.get_smiles_list(
+        solvent_primary_keys_ls
+    )
+    return solvent_dict
+
+
+def get_product_data() -> Dict:
+    """
+    Get product data from the request
+    Returns:
+         Dictionary with product data stored as a list with one entry per product or a single value.
+    """
+    product_data_keys = [
+        "productTableNumbers",
+        "products",
+        "productMasses",
+        "roundedProductMasses",
+        "productMolecularWeights",
+        "productHazards",
+        "productPhysicalForms",
+    ]
+    product_dict = get_request_data_from_keys(product_data_keys)
+    services.hazard_code.get_multiple_compounds_data(product_dict, "product")
+
+    # Primary keys and SMILES
+    # Joining primary keys and getting smiles
+    product_primary_keys_ls = auxiliary.get_data("productPrimaryKeys")
+    product_dict["product_primary_keys_str"] = ", ".join(product_primary_keys_ls)
+    product_dict["product_smiles_ls"] = services.all_compounds.get_smiles_list(
+        product_primary_keys_ls
+    )
+    # Table numbers
+    product_table_numbers = auxiliary.get_data("productTableNumbers")
+    product_dict["product_table_numbers"] = list(filter(None, product_table_numbers))
+    product_dict["main_product_table_number"] = int(
+        request.form["mainProductTableNumber"]
+    )
+
+    for idx, product_table_num in enumerate(product_dict["product_table_numbers"]):
+        if int(product_table_num) == product_dict["main_product_table_number"]:
+            product_dict["main_product_index"] = idx
+            break
+    return product_dict
+
+
+def get_unit_data() -> Dict:
+    """
+    Gets a dictionary for the unit data from the reaction table.
+    Returns:
+        dictionary with unit data
+    """
+    # Gets selected units from the reaction table
+    unit_dict = {
+        "amount_unit": str(request.form["amountUnit"]),
```

Comment:
:seedling: I'd begin to inject the request.form as a parameter to the function, and generally move away from the global variable request where you can in the future 
It's modular, explicit, and easier to test / understand.
(and nice job on more tests, I still feel bad for breaking them all)

Reply:
Cheers Andy, I have implemented in this PR

Reply:
And no worries on the tests they're coming back and they're definitely easier to run/develop/maintain now!